### PR TITLE
827 Refactor load balancing to allow running strategies abstractly from inputs and effects

### LINKED ISSUE #827: Refactor LB to make strategies closer to pure functions from stats -> decisions
**Is your feature request related to a problem? Please describe.**

It's currently impossible to run an LB strategy just to see what it would do, because they indirectly call straight out to the function that would actually migrate objects according to their decisions. This means we can't run multiple strategies and pick the one with the best result, or run a strategy with different sets of tuning parameters, or with differently modeled projections of expected object loads.

**What Needs to be Done?**

- [x] Define a structure for a list of migration decisions from an LB strategy invocation
- [x] Convert LB manager and strategies to operate in terms of that structure
- [ ] Change the LB strategy interface to take statistics as arguments, rather than grabbing them from `theProcStats()`, and the caller to pass them

**Describe potential solution outcome**

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #885: vtAssert fails in BaseLB about migrateObjectTo
**Describe the bug**
In `22bf5db3359b8a9f5f6c14ad9ba9d0ac598db128`, an assertion was added to develop that if `migrateObjectTo` is called a current migration epoch is open. However, this assertion fails inside BaseLB! This is the case where the migration needs to transferred to another node:

```C++
void BaseLB::transferMigrations(TransferMsg<TransferVecType>* msg) {
  auto const& migrate_list = msg->getTransfer();
  for (auto&& elm : migrate_list) {
    auto obj_id  = std::get<0>(elm);
    auto to_node = std::get<1>(elm);
    vtAssert(theProcStats()->hasObjectToMigrate(obj_id), "Must have object");
    migrateObjectTo(obj_id, to_node);
  }
}
```

This assertion is causing all my runs with LB to fail on develop.

### PR DESCRIPTION
**Pull request by PhilMiller**
Fixes #827
This should also: Fixes #885

### CODE REVIEW

**Comment by PhilMiller** [comment]
@lifflander This is not in any sense done yet, but please have a look to make sure it's not going in a terrible direction.

**Comment by lifflander** [comment]
> @lifflander This is not in any sense done yet, but please have a look to make sure it's not going in a terrible direction.

This looks like a great start to me! I was thinking if we are refactoring BaseLB (out of the hierarchy), we might want to split out some of the purely functional LB helpers into a specific helper class.

```
  ObjBinType histogramSample(LoadType const& load) const;
  LoadType loadMilli(LoadType const& load) const;
  int32_t getBinSize() const { return bin_size_; }
  NodeType objGetNode(ObjIDType const id) const;
```

**Comment by PhilMiller** [comment]
I don't see a particular need to eliminate `BaseLB` from the hierarchy entirely - I think a common base class that can be used as a pointer type to any strategy is valuable, if nothing else.

Those helper functions you mentioned may not be universally used, but they don't really get in the way of anything else we want to do to advance the LB capabilities of the system.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Comment by PhilMiller** [comment]
I'd like to get this reviewed and merged, and I'll move forward with the ID cleanup in a follow-on issue and PR. I think this will be easier to review separate from that.

**Comment by PhilMiller** [comment]
That'll be #868

**Comment by lifflander** [comment]
@PhilMiller Can you label your commits with a prefix of the issue number #827 so we can refer to them easily?

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

File: src/vt/vrt/collection/balance/lb_invoke/lb_manager.cc [hunk 1]
```diff
@@ -126,9 +126,41 @@ template <typename LB>
...
+
+  EpochType balance_epoch = theTerm()->makeEpochCollective("LBManager::balance_epoch");
+  theMsg()->pushEpoch(balance_epoch);
+  strat->startLB(phase, base_proxy, theProcStats()->getProcLoad(phase), theProcStats()->getProcComm(phase));
+  theMsg()->popEpoch(balance_epoch);
+  theTerm()->finishedEpoch(balance_epoch);
+
+  EpochType migrate_epoch = theTerm()->makeEpochCollective("LBManager::migrate_epoch");
+
+  theTerm()->addAction(balance_epoch,
```
[the remark this answers is not in the document]
Comment (lifflander) [reply src/vt/vrt/collection/balance/lb_invoke/lb_manager.cc:142]: Fixed

File: src/vt/vrt/collection/balance/baselb/baselb.cc [hunk 2]
```diff
@@ -163,63 +168,59 @@ void BaseLB::statsHandler(StatsMsgType* msg) {
...
-}
+  for (auto&& elm : transfers) {
+    auto obj_id = std::get<0>(elm);
+    auto to = std::get<1>(elm);
+    auto from = objGetNode(obj_id);
+
+    if (from != to) {
+      bool has_object = theProcStats()->hasObjectToMigrate(obj_id);
+
+      debug_print(
```
[the remark this answers is not in the document]
Comment (lifflander) [reply src/vt/vrt/collection/balance/baselb/baselb.cc:196]: Fixed

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

File: src/vt/vrt/collection/balance/lb_invoke/lb_manager.cc [hunk 3]
```diff
@@ -126,9 +126,41 @@ template <typename LB>
...
 LBManager::makeLB(MsgSharedPtr<StartLBMsg> msg) {
   auto proxy = theObjGroup()->makeCollective<LB>();
-  proxy.get()->init(proxy);
+  auto strat = proxy.get();
+  strat->init(proxy);
   auto base_proxy = proxy.template registerBaseCollective<lb::BaseLB>();
-  proxy.get()->startLBHandler(msg.get(), base_proxy);
+  auto phase = msg->getPhase();
+
+  EpochType balance_epoch = theTerm()->makeEpochCollective("LBManager::balance_epoch");
```
[the remark this answers is not in the document]
Comment (PhilMiller) [reply src/vt/vrt/collection/balance/lb_invoke/lb_manager.cc:134]: Strongly agreed.

File: src/vt/vrt/collection/balance/lb_invoke/lb_manager.cc [hunk 4]
```diff
@@ -126,9 +126,41 @@ template <typename LB>
...
+				     lb, node,
+				     "LBManager: starting migrations\n"
+				     );
+			 theMsg()->pushEpoch(migrate_epoch);
+			 strat->applyMigrations(strat->getTransfers());
+			 theMsg()->popEpoch(migrate_epoch);
+			 theTerm()->finishedEpoch(migrate_epoch);
+		       });
+
+  theTerm()->addAction(migrate_epoch,
```
[the remark this answers is not in the document]
Comment (PhilMiller) [reply src/vt/vrt/collection/balance/lb_invoke/lb_manager.cc]: We don't. The continuation the collection elements passed when they said they were ready for LB won't be called until it's done, though.

Do I need to look further up the stack, to see that something is actually going to drive this forward?

[the remark this answers is not in the document]
Comment (lifflander) [reply src/vt/vrt/collection/balance/lb_invoke/lb_manager.cc]: Right, `LBManager::sysLB` is called by `ElementStats::syncNextPhase` after a reduction across the collection. `sysLB` is the function that calls `LBManager::collectiveImpl`. I don't see anything that will ensure this happens before control returns back to the user code.

I think we need to wait on `migrate_epoch`. What do you think?

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

File: src/vt/vrt/collection/manager.impl.h [hunk 5]
```diff
@@ -2978,9 +2978,11 @@ void CollectionManager::nextPhase(
     );
   #endif
 
-  theCollection()->broadcastMsg<MsgType,ElementStats::syncNextPhase<ColT>>(
-    proxy, msg.get(), instrument
-  );
+  runInEpochCollective([&] {
```
Comment (lifflander) [inline src/vt/vrt/collection/manager.impl.h:2981]: This is going to create a epoch for each collection---not what we want. `nextPhase` is currently called on a per-collection basis.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/vt/vrt/collection/balance/zoltanlb/zoltanlb.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5707118&bid=18679258)

**Pull request merged**